src/symbol/quads.js

```diff
@@ -265,62 +302,123 @@ export function getGlyphQuads(anchor: Anchor,
+                //   |     |
+                // bl ----- br
+                tl = new Point(-halfAdvance + builtInOffset[0], builtInOffset[1]);
+                tl._rotateAround(verticalRotation, center)._add(verticalOffsetCorrection);
+
+                // Relative position after rotating
+                // tr ----- br
+                //   |     |
+                //   |     |
+                // tl ----- bl
+                // After rotation, glyph lies on the horizontal midline.
+                // Shift back to tl's original x coordinate before rotation by applying 'xOffsetCorrection'.
+                const offsetCorrection = new Point(-yShift + halfAdvance, 0);
+                tl._add(offsetCorrection);
+
+                // Add padding for y coordinate's justification
+                tl.y -= (metrics.left - rectBuffer) * positionedGlyph.scale;
+
+                // Adjust x coordinate according to glyph bitmap's height and the vectical advance
+                const verticalAdvance = positionedGlyph.imageName ? metrics.advance * positionedGlyph.scale :
+                    ONE_EM * positionedGlyph.scale;
+                // Check wether the glyph is generated from server side or locally
+                const chr = String.fromCharCode(positionedGlyph.glyph);
+                if (isVerticalClosePunctuation(chr)) {
+                    // Place vertical punctuation in right place, pull down 1 pixel's space for close punctuations
+                    tl.x += (-rectBuffer + 1) * positionedGlyph.scale;
+                } else if (isVerticalOpenPunctuation(chr)) {
+                    const xOffset = verticalAdvance - metrics.height * positionedGlyph.scale;
+                    // Place vertical punctuation in right place, pull up 1 pixel's space for open punctuations
+                    tl.x += xOffset + (-rectBuffer - 1) * positionedGlyph.scale;
+                } else if (!positionedGlyph.imageName &&
+                           ((metrics.width + rectBuffer * 2) !== rect.w || metrics.height + rectBuffer * 2 !== rect.h)) {
+                    // Locally generated glyphs' bitmap do not have exact 'rectBuffer' padded around the glyphs,
+                    // but the original tl do have distance of rectBuffer padded to the top of the glyph.
+                    const perfectPaddedHeight = (metrics.height + rectBuffer * 2) * positionedGlyph.scale;
+                    const delta = verticalAdvance - perfectPaddedHeight;
+                    tl.x += delta / 2;
+                } else {
+                    // Place the glyph bitmap right in the center of the 24x24 point boxes
+                    const delta = verticalAdvance - paddedHeight;
+                    tl.x += delta / 2;
+                }
+                // Calculate other three points
+                tr = tl.add(new Point(0, -paddedWidth));
+                bl = tl.add(new Point(paddedHeight, 0));
+                br = tl .add(new Point(paddedHeight, -paddedWidth));
```

Comment:
Nit: let's avoid `Point` allocations where not needed — e.g. doing `tr = new Point(...)` directly would make it one instead of two.